HyperVExtension/src/HyperVExtension/Models/HyperVVirtualMachine.cs

```diff
@@ -1,657 +1,798 @@
+
+                                // Create SDK's result. Set Completed status and event.
+                                operation.SetState(ConfigurationSetState.Completed, null, configureResponse.ApplyConfigurationResult, null);
+
+                                // Receiving ConfigureResponse means operation has completed. We don't expect anymore responses.
+                                isCompleted = true;
+                                break;
+                            }
+                            else if (response is ConfigureProgressResponse configureProgressResponse)
+                            {
+                                LogApplyConfigurationProgress(configureProgressResponse.ProgressData);
+
+                                // Create SDK's result. Set Completed status and event.
+                                operation.SetState(ConfigurationSetState.InProgress, configureProgressResponse.ProgressData, null, null);
+                            }
+                            else
+                            {
+                                // Unexpected (error) response. Log it and return error. Not much we can do here.
+                                Logging.Logger()?.ReportError(
+                                    $"Unexpected response while applying configuration on {DateTime.Now}: " +
+                                    $"responseId: {response.RequestId}, responseType: {response.ResponseType}, " +
+                                    $"VM details: {this}");
+                            }
+                        }
+                    }
+                }
+                catch (Exception ex)
+                {
+                    Logging.Logger()?.ReportError($"Failed to apply configuration on {DateTime.Now}: VM details: {this}", ex);
+                    operation.SetState(ConfigurationSetState.Completed, null, new HostGuestCommunication.ApplyConfigurationResult(ex.HResult, ex.Message), null);
+                }
+            });
+
+            return operation;
+        }
+        catch (Exception ex)
+        {
+            Logging.Logger()?.ReportError($"Failed to apply configuration on {DateTime.Now}: VM details: {this}", ex);
+            return new ApplyConfigurationOperation(this, ex);
+        }
+    }
+
+    private bool DeployDevSetupAgent(ApplyConfigurationOperation operation)
+    {
+        var powerShell = _host.GetService<IPowerShellService>();
+        var credentialsAdaptiveCardSession = new VmCredentialAdaptiveCardSession(operation);
+
+        operation.SetState(ConfigurationSetState.WaitingForAdminUserLogon, null, null, credentialsAdaptiveCardSession);
+
+        (var userName, var password) = credentialsAdaptiveCardSession.WaitForCredentials();
+
+        if ((userName != null) && (password != null))
+        {
+            var deploymentHelper = new DevSetupAgentDeploymentHelper(powerShell, Id);

```

Comment:
I think `DevSetupAgentDeploymentHelper` should actually be created as a transient service that the `HyperVVirtualMachine` takes in its constructor. Would also be good if it implemented an interface or an abstract class so we can unit test this easier. Then we just pass the Id of the VM to DevSetupAgentDeploymentHelper's function methods. We can make that change in the future. IHost wouldn't be needed in that case. I plan to create a factory delegate method for the HyperVVirtualMachine and register it to the host container, which will inject any registered service in its constructor, and anything else it needs like the PSObject. But I will do that in the future.

Reply:
Yes we can do that. However, I am not sure about passing it in the constructor. DevSetupAgentDeploymentHelper is only used sometimes, when setup is needed and for the duration of setup. Creating it for every instance of HyperVVirtualMachine is a vaste of resources.

Reply:
fair enough, we can think about that later on. I don't think the constructor arguments for HyperVVirtualMachine  will get any bigger in the meantime